node/cn/peer_set_test.go

```diff
@@ -0,0 +1,654 @@
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the klaytn library. If not, see <http://www.gnu.org/licenses/>.
+
+package cn
+
+import (
+	"github.com/golang/mock/gomock"
+	"github.com/klaytn/klaytn/blockchain/types"
+	"github.com/klaytn/klaytn/common"
+	"github.com/klaytn/klaytn/networks/p2p"
+	"github.com/klaytn/klaytn/node"
+	"github.com/stretchr/testify/assert"
+	"math/big"
+	"testing"
+)
+
+func TestPeerSet_Register(t *testing.T) {
+	peerSet := newPeerSet()
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	cnPeer := NewMockPeer(mockCtrl)
+	cnPeer.EXPECT().ConnType().Return(p2p.ConnType(node.CONSENSUSNODE)).Times(1)
+	cnPeer.EXPECT().GetAddr().Return(addr1).Times(3)
+	cnPeer.EXPECT().GetID().Return(nodeID1.String()).Times(3)
+	cnPeer.EXPECT().Broadcast().AnyTimes()
+
+	pnPeer := NewMockPeer(mockCtrl)
+	pnPeer.EXPECT().ConnType().Return(p2p.ConnType(node.PROXYNODE)).Times(1)
+	pnPeer.EXPECT().GetAddr().Return(addr2).Times(3)
+	pnPeer.EXPECT().GetID().Return(nodeID2.String()).Times(3)
+	pnPeer.EXPECT().Broadcast().AnyTimes()
+
+	enPeer := NewMockPeer(mockCtrl)
+	enPeer.EXPECT().ConnType().Return(p2p.ConnType(node.ENDPOINTNODE)).Times(1)
+	enPeer.EXPECT().GetAddr().Return(addr3).Times(3)
+	enPeer.EXPECT().GetID().Return(nodeID3.String()).Times(3)
+	enPeer.EXPECT().Broadcast().AnyTimes()
+
+	assert.NoError(t, peerSet.Register(cnPeer))
+	assert.NoError(t, peerSet.Register(pnPeer))
+	assert.NoError(t, peerSet.Register(enPeer))
+
+	assert.Equal(t, errAlreadyRegistered, peerSet.Register(cnPeer))
+	assert.Equal(t, errAlreadyRegistered, peerSet.Register(pnPeer))
+	assert.Equal(t, errAlreadyRegistered, peerSet.Register(enPeer))
+
+	peerSet.closed = true
+	assert.Equal(t, errClosed, peerSet.Register(cnPeer))
+	assert.Equal(t, errClosed, peerSet.Register(pnPeer))
+	assert.Equal(t, errClosed, peerSet.Register(enPeer))
+}
+
+func TestPeerSet_Unregister(t *testing.T) {
+	peerSet := newPeerSet()
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	cnPeer := NewMockPeer(mockCtrl)
+	cnPeer.EXPECT().ConnType().Return(p2p.ConnType(node.CONSENSUSNODE)).Times(2)
+	cnPeer.EXPECT().GetAddr().Return(addr1).Times(4)
+	cnPeer.EXPECT().GetID().Return(nodeID1.String()).Times(2)
+	cnPeer.EXPECT().Broadcast().AnyTimes()
+	cnPeer.EXPECT().Close().Times(1)
+
+	pnPeer := NewMockPeer(mockCtrl)
+	pnPeer.EXPECT().ConnType().Return(p2p.ConnType(node.PROXYNODE)).Times(2)
+	pnPeer.EXPECT().GetAddr().Return(addr2).Times(4)
+	pnPeer.EXPECT().GetID().Return(nodeID2.String()).Times(2)
+	pnPeer.EXPECT().Broadcast().AnyTimes()
+	pnPeer.EXPECT().Close().Times(1)
+
+	enPeer := NewMockPeer(mockCtrl)
+	enPeer.EXPECT().ConnType().Return(p2p.ConnType(node.ENDPOINTNODE)).Times(2)
+	enPeer.EXPECT().GetAddr().Return(addr3).Times(4)
+	enPeer.EXPECT().GetID().Return(nodeID3.String()).Times(2)
+	enPeer.EXPECT().Broadcast().AnyTimes()
+	enPeer.EXPECT().Close().Times(1)
+
```

Comment:
How about ensuring peer un-registration by checking the decrement of `len(ps.peers)`?